go/types/list.go

```diff
@@ -108,11 +106,41 @@ func (l List) Iter(f listIterFunc) {
 	})
 }
 
-type listIterAllFunc func(v Value, index uint64)
+func (l List) IterRange(f func(v Value, idx uint64), startIdx, endIdx uint64) {
```

Comment:
This is basically `copyReadAhead`, which makes sense, because "copy read ahead" is basically "get my all values in a range". It seems like it'd be relatively easy to share the bulk of this code; whereas `copyReadAhead` returns `[]Value`, this function runs a callback.

Though honestly it's almost like we should have `Slice` instead of `IterRange`, because returning a slice is more direct and flexible (for the caller), and it more accurately represents how this is implemented.

... Unless you wanted to evolve this function later to divide up the iteration range if it's large (for parallelism or to save memory), i.e. refactor `IterAll` into `IterRange` and implement the former with `IterRange(0, l.Len())`.

Reply:
Or `GoSlice`, since we may want a prollytree-level `Slice` down the road.

Fun idea, but probably unnecessary at this point:
```go
func (l List) GoSlice(range ...uint64) []Value {
  var from, to uint64

  if len(range) > 0 {
    from = range[0]
    if len(range) > 1 {
      to = range[1]
      if len(range) > 2 {
        panic("too many arguments to GoRange")
      }
    } else {
      to = l.Len()
    }
  }
}
```

Reply:
^ updated this